rfcs/XXXX-gating/README.md

```diff
@@ -0,0 +1,301 @@
+
+```yaml
+apiVersion: gating.toolkit.fluxcd.io/v1alpha1
+kind: Gate
+metadata:
+  name: sre-approval
+  namespace: flux-system
+status:
+  requestedAt: "2021-03-26T10:00:00Z"
+  resetToDefaultAt: "2021-03-26T11:00:00Z"
+  conditions:
+    - lastTransitionTime: "2021-03-26T10:00:00Z"
+      message: "Gate scheduled for closing at 2021-03-26T11:00:00Z"
+      reason: ReconciliationSucceeded
+      status: "True"
+      type: Opened
+```
+
+While the gate is opened, all the objects that reference it are approved to reconcile at their configured interval.
+
+The SRE can decide to close the gate ahead of its schedule with:
+
+```sh
+kubectl -n flux-system annotate --overwrite gate/sre-approval \
+close.gate.fluxcd.io/requestedAt="$(date -u +"%Y-%m-%dT%H:%M:%SZ")"
+```
+
+The `gating-controller` extracts the ISO8601 date from the `close.gate` annotation value,
+compares it with the `open.gate` & `requestedAt` date and closes the gate:
+
+```yaml
+apiVersion: gating.toolkit.fluxcd.io/v1alpha1
+kind: Gate
+metadata:
+  name: sre-approval
+  namespace: flux-system
+status:
+  requestedAt: "2021-03-26T10:10:00Z"
+  resetToDefaultAt: "2021-03-26T10:10:00Z"
+  conditions:
+    - lastTransitionTime: "2021-03-26T10:10:00Z"
+      message: "Gate close requested"
+      reason: ReconciliationSucceeded
+      status: "False"
+      type: Opened
+```
+
+The objects that are referencing this gate, will finish their ongoing reconciliation (if any) then pause.
+
+> As a member of the SRE team, I want to block deployments in a particular time window.
+
+To enforce a maintenance window of 24 hours, you can define a `Gate` that's opened by default:
+
+```yaml
+apiVersion: gating.toolkit.fluxcd.io/v1alpha1
+kind: Gate
+metadata:
+  name: maintenance
+  namespace: flux-system
+spec:
+  interval: 30s
+  default: opened
+  window: 24h
+```
+
+To start the maintenance window you can annotate the gate with:
+
+```sh
+kubectl -n flux-system annotate --overwrite gate/maintenance \
+close.gate.fluxcd.io/requestedAt="$(date -u +"%Y-%m-%dT%H:%M:%SZ")"
+```
+
+The `gating-controller` extracts the ISO8601 date from the `close.gate`
+annotation value and closes the gate for the specified window:
+
+```yaml
+apiVersion: gating.toolkit.fluxcd.io/v1alpha1
+kind: Gate
+metadata:
+  name: maintenance
+  namespace: flux-system
+status:
+  requestedAt: "2021-03-26T10:00:00Z"
+  resetToDefaultAt: "2021-03-27T10:00:00Z"
+  conditions:
+    - lastTransitionTime: "2021-03-26T10:00:00Z"
+      message: "Gate scheduled for opening at 2021-03-27T11:00:00Z"
+      reason: ReconciliationSucceeded
+      status: "False"
+      type: Opened
+```
+
+You could also schedule "No Deploy Fridays" with a CronJob that closes the `maintenance` gate at `0 0 * * FRI`.
+
+#### Story 2
+
+> As a member of the SRE team, I want existing deployments to still be
+> reconciled during a change freeze.
+
+Gates can be used to block Flux sources from being refreshed, resulting in Flux
```

Comment:
To clarify that `block Flux sources from being refreshed` in this case means the gate prevent the source controller from making the new assets available, rather than to suspend the polling in the source controller?

It is desirable for the source to still be able to detect changes, such that in-cluster logic that is tied to the gating mechanisms can be notified of available changes.

Reply:
If it stopped polling, it would avoid all the resource consumption (CPU, memory, Network and storage) for an operation that would not cause any side effect to the cluster, which would stream line the controllers to better handle reconciliations that were not gated.

@hassenius is there any specific scenario you have in mind?